### PR TITLE
refactor(meta): reduce meta store write in hummock service

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Refactor hummock service:
- Remove 2 persistent states: stale_sstables and sstable_id_infos.
- Replace hummock_versions with checkpoint_version.
- Simplify vacuum. 

TODO:
- (next PR) GC orphan SSTs in object store. Also check SST validity in commit_epoch and report_compact_task.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/singularity-data/risingwave/issues/4037

### CODE REVIEW

File: src/meta/src/hummock/hummock_manager.rs [hunk 1]
```diff
@@ -177,27 +174,58 @@ macro_rules! start_measure_real_process_timer {
...
+    pub fn extend_ssts_to_delete_from_deltas(
+        &mut self,
+        delta_range: impl RangeBounds<HummockVersionId>,
+    ) {
+        for (_, delta) in self.hummock_version_deltas.range(delta_range) {
+            let mut no_sst_to_delete = true;
+            for (_, level_deltas) in delta.level_deltas.iter() {
+                for level_delta in level_deltas.level_deltas.iter() {
+                    for sst_id in level_delta.removed_table_ids.iter() {
+                        assert!(self.ssts_to_delete.insert(*sst_id, delta.id).is_none());
```
Comment (soundOfDestiny) [inline src/meta/src/hummock/hummock_manager.rs:217]: I don't like side-effects in `assert`. can we split this?

File: src/meta/src/hummock/hummock_manager.rs [hunk 2]
```diff
@@ -1196,180 +1121,36 @@ where
...
-                let sst_id = ssts_to_delete.id[idx];
-                if !ssts_in_use.contains(&sst_id) && let Some(mut sst_id_info) = sstable_id_infos.get_mut(sst_id) {
-                    sst_id_info.meta_delete_timestamp = sstable_id_info::get_timestamp_now();
-                    // We don't want to repetitively set the delete timestamp of these that have been set,
-                    // so we remove these ones.
-                    ssts_to_delete.id.swap_remove(idx);
-                }
-            }
+        let min_pinned_version_id = versioning_guard.min_pinned_version_id();
+        if min_pinned_version_id <= versioning_guard.checkpoint_version.id {
```
Comment (soundOfDestiny) [inline src/meta/src/hummock/hummock_manager.rs:1130]: I think if `min_pinned_version_id` == `u64::MAX`, we should also return here

Comment (soundOfDestiny) [reply src/meta/src/hummock/hummock_manager.rs:1130]: > I think if `min_pinned_version_id` == `u64::MAX`, we should also return here

oh, I'm incorrect.

File: src/meta/src/hummock/hummock_manager.rs [hunk 3]
```diff
@@ -1196,180 +1121,36 @@ where
...
+        for (_, version_delta) in versioning
+            .hummock_version_deltas
+            .range((Excluded(old_checkpoint_id), Included(new_checkpoint_id)))
         {
-            drop(versioning_guard);
-            self.check_state_consistency().await;
+            checkpoint.apply_version_delta(version_delta);
         }
-
+        commit_multi_var!(self, None, checkpoint)?;
```
Comment (soundOfDestiny) [inline src/meta/src/hummock/hummock_manager.rs:1143]: if min_pinned_version_id == u64::MAX, it's possible that we do empty commit here. we can avoid this.

**Review by soundOfDestiny — APPROVED** [review]
lgtm

File: src/meta/src/hummock/hummock_manager.rs (hunk 3 above)
Comment (zwang28) [reply src/meta/src/hummock/hummock_manager.rs:1163]: fixed

**Comment by zwang28** [comment]
@lmatz FYI the optimization from https://github.com/singularity-data/risingwave/pull/2315 is reverted in this PR, because we now only maintains min_pinned_version for each node, rather than the accurate pinned versions. As a result we cannot identify SSTs qualified for deletion in versions whose id > min_pinned_version_id any more.

**Comment by lmatz** [comment]
> @lmatz FYI the optimization from https://github.com/singularity-data/risingwave/pull/2315 is reverted in this PR, because we now only maintains min_pinned_version for each node (since https://github.com/singularity-data/risingwave/pull/3850), rather than the accurate pinned versions. As a result we cannot identify SSTs qualified for deletion in versions whose id > min_pinned_version_id any more.
> 
> 

Thanks!Good to know👍

**Pull request merged**